### PR TITLE
System: Generalise Options

### PR DESCRIPTION
**Pull request by ao-david**
This change generalises the options interface so that methods such as WithWifi can be reused for functions other than NewRigSystem.

The idea is that WithWifi can also be used in a NewPeripheral method in the future without needing to give it a different name, or putting it in another package.

Suggestions are appreciated, I am not sure if this is the best way of doing this.

### CODE REVIEW

File: system/options.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/ausocean/cloud/model"
+)
+
+// Option represents functional options that can be passed to NewRigSystem.
+type Option func(any) error
+
+// variableHolder is an interface for types which can add variables.
+type variableHolder interface {
+	AddVariables(variables ...*model.Variable)
```
[the remark this answers is not in the document]
Comment (ao-david) [reply system/options.go:36]: Are you suggesting renaming, or adding the function set?

[the remark this answers is not in the document]
Comment (saxon-milton) [reply system/options.go:36]: I think rename. IMO there's a difference between "Add" and "Set". "Add" is like appending to a list and "Set" is like setting/replacing a single thing (which could be a whole list).

[the remark this answers is not in the document]
Comment (ao-david) [reply system/options.go:36]: I mean the specific implementation of this function does append the variables to the device

```golang
// addVariables adds variables to the rig system controller. This
// implements the variableHolder interface.
func (sys *RigSystem) AddVariables(variables ...*model.Variable) {
	sys.Variables = append(sys.Variables, variables...)
}
```

[the remark this answers is not in the document]
Comment (saxon-milton) [reply system/options.go:36]: Yeah sorry, I think I got my wires crossed. I reckon I was looking at AddWifi.

File: system/options.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+			return fmt.Errorf("%v does not implement variableHolder interface", reflect.TypeOf(v).String())
+		}
+		vh.AddVariables(variables...)
+		return nil
+	}
+}
+
+// wifiHolder is an interface for types which can add wifi.
+type wifiHolder interface {
+	AddWifi(ssid, pass string)
```
[the remark this answers is not in the document]
Comment (trekhopton) [reply system/options.go:53]: I agree, this should be _set_.

File: system/options.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+			return nil
+		}
+		wh.AddWifi(ssid, pass)
+		return nil
+	}
+}
+
+// locationHolder is an interface for types which can set their location.
+type locationHolder interface {
+	AddLocation(lat, long float64)
```
Comment (trekhopton) [inline system/options.go:74]: *SetLocation

**Review by saxon-milton — APPROVED** [review]

**Pull request merged**